apps/src/lib/tools/jsdebugger/redux.js

```diff
@@ -0,0 +1,261 @@
+import * as Immutable from 'immutable';
+import Observer from '../../../Observer';
+import {update as updateWatchExpressions} from '../../../redux/watchedExpressions';
+import CommandHistory from './CommandHistory';
+import JSInterpreter from '../../../JSInterpreter';
+import watchedExpressions from '@cdo/apps/redux/watchedExpressions';
+
+const WATCH_TIMER_PERIOD = 250;
+const INITIALIZE = 'jsdebugger/INITIALIZE';
+const ATTACH = 'jsdebugger/ATTACH';
+const DETACH = 'jsdebugger/DETACH';
+const MIRROR_INTERPRETER_STATE = 'jsdebugger/MIRROR_INTERPRETER_STATE';
+const APPEND_LOG = 'jsdebugger/APPEND_LOG';
+const CLEAR_LOG = 'jsdebugger/CLEAR_LOG';
+
+// State model
+const JSDebuggerState = Immutable.Record({
+  jsInterpreter: null,
+  runApp: null,
+  observer: null,
+  watchIntervalId: null,
+  commandHistory: null,
+  logOutput: '',
+  jsInterpreterNextStep: null,
+  jsInterpreterPaused: false,
+});
+
+export function getRoot(state) {
+  return state.jsdebugger;
+}
+
+export function getCommandHistory(state) {
+  return getRoot(state).commandHistory;
+}
+
+export function getJSInterpreter(state) {
+  return getRoot(state).jsInterpreter;
+}
+
+function getRunApp(state) {
+  return getRoot(state).runApp;
+}
+
+export function isPaused(state) {
+  return getRoot(state).jsInterpreterPaused;
+}
+
+function getObserver(state) {
+  return getRoot(state).observer;
+}
+
+function getWatchIntervalId(state) {
+  return getRoot(state).watchIntervalId;
+}
+
+export function isAttached(state) {
+  return !!getJSInterpreter(state);
+}
+
+export function canRunNext(state) {
+  return (
+    isPaused(state) &&
+    getRoot(state).jsInterpreterNextStep === JSInterpreter.StepType.RUN
+  );
+}
+
+export function getLogOutput(state) {
+  return getRoot(state).logOutput;
+}
+
+export const selectors = {
+  getRoot,
+  getCommandHistory,
+  getJSInterpreter,
+  isPaused,
+  isAttached,
+  canRunNext,
+  getLogOutput,
+};
+
+// actions
+export function initialize({runApp}) {
+  return {type: INITIALIZE, runApp};
+}
+
+export function appendLog(output) {
+  return {type: APPEND_LOG, output};
+}
+
+export function clearLog() {
+  return {type: CLEAR_LOG};
+}
+
+function mirrorInterpreterState(jsInterpreter) {
+  return {type: MIRROR_INTERPRETER_STATE, jsInterpreter};
```

Comment:
We've talked for a while about giving the interpreter a real debugger protocol; maybe some sane subset of [Chrome Debugging Protocol](https://developer.chrome.com/devtools/docs/debugger-protocol), or just its [JavaScript debugger section](https://chromedevtools.github.io/debugger-protocol-viewer/1-2/Debugger/). Even in that world, though, you're going to be responding to interpreter events, and making queries to it, not directly mirroring its internal state, right?  In view of that possible future, is this the right way to go, or should the debugger hold a copy of specific state it cares about and have a clear-cut event interface with the interpreter to update its internal model?